### PR TITLE
Implement firebase-database Kotlin extension.

### PR DESCRIPTION
**Pull request by rlazo**

### CODE REVIEW

File: firebase-database/ktx/src/main/kotlin/com/google/firebase/database/ktx/Database.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+FirebaseDatabase.getInstance(app, url)
+
+/**
+ * Returns the content of the DataSnapshot converted to a POJO.
+ *
+ * Supports generics like List<> or Map<>. Use @JvmSuppressWildcards to force the compiler to
+ * use the type `T`, and not `? extends T`.
+ */
+inline fun <reified T> DataSnapshot.getValue(): T? {
+    var genericTypeIndicator = object : GenericTypeIndicator<T>() {}
```
Comment (rosariopfernandes) [inline firebase-database/ktx/src/main/kotlin/com/google/firebase/database/ktx/Database.kt]: nit: maybe use `val` instead of `var` here.

Comment (rlazo) [reply firebase-database/ktx/src/main/kotlin/com/google/firebase/database/ktx/Database.kt]: Good catch!. Fixed

File: firebase-database/ktx/src/main/kotlin/com/google/firebase/database/ktx/Database.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+FirebaseDatabase.getInstance(app, url)
+
+/**
+ * Returns the content of the DataSnapshot converted to a POJO.
+ *
+ * Supports generics like List<> or Map<>. Use @JvmSuppressWildcards to force the compiler to
+ * use the type `T`, and not `? extends T`.
+ */
+inline fun <reified T> DataSnapshot.getValue(): T? {
+    val genericTypeIndicator = object : GenericTypeIndicator<T>() {}
```
Comment (vkryachko) [inline firebase-database/ktx/src/main/kotlin/com/google/firebase/database/ktx/Database.kt]: worth inlining into `getValue` call?

**Comment by msasikanth** [comment]
Also add a extension supporting Kotlin coroutines?

```
suspend fun <T : Any> Query.await(): T? {
    return suspendCancellableCoroutine { continuation ->
        val eventListener = object : ValueEventListener {
            override fun onDataChange(dataSnap: DataSnapshot) {
                val type = object : GenericTypeIndicator<T>() {}
                continuation.resume(dataSnap.getValue(type))
            }

            override fun onCancelled(dbError: DatabaseError) {
                continuation.resumeWithException(dbError.toException())
            }
        }
        continuation.invokeOnCancellation {
            removeEventListener(eventListener)
        }
        addListenerForSingleValueEvent(eventListener)
    }
}
```

File: firebase-database/ktx/src/main/kotlin/com/google/firebase/database/ktx/Database.kt (hunk 2 above)
Comment (rlazo) [reply firebase-database/ktx/src/main/kotlin/com/google/firebase/database/ktx/Database.kt]: Good catch, fixed.

**Comment by msasikanth** [comment]
> Also add an extension supporting Kotlin coroutines? I can provide a PR for this separately.
> 
> ```
> suspend fun <T : Any> Query.await(): T? {
>     return suspendCancellableCoroutine { continuation ->
>         val eventListener = object : ValueEventListener {
>             override fun onDataChange(dataSnap: DataSnapshot) {
>                 val type = object : GenericTypeIndicator<T>() {}
>                 continuation.resume(dataSnap.getValue(type))
>             }
> 
>             override fun onCancelled(dbError: DatabaseError) {
>                 continuation.resumeWithException(dbError.toException())
>             }
>         }
>         continuation.invokeOnCancellation {
>             removeEventListener(eventListener)
>         }
>         addListenerForSingleValueEvent(eventListener)
>     }
> }
> ```

This extension may not work as expected since type is being erased at runtime. In order to fix that class type should be passed as parameter of the function or make the function inline and use reified.

**Comment by rlazo** [comment]
> > Also add an extension supporting Kotlin coroutines? I can provide a PR for this separately.
> > ```
> > suspend fun <T : Any> Query.await(): T? {
> >     return suspendCancellableCoroutine { continuation ->
> >         val eventListener = object : ValueEventListener {
> >             override fun onDataChange(dataSnap: DataSnapshot) {
> >                 val type = object : GenericTypeIndicator<T>() {}
> >                 continuation.resume(dataSnap.getValue(type))
> >             }
> > 
> >             override fun onCancelled(dbError: DatabaseError) {
> >                 continuation.resumeWithException(dbError.toException())
> >             }
> >         }
> >         continuation.invokeOnCancellation {
> >             removeEventListener(eventListener)
> >         }
> >         addListenerForSingleValueEvent(eventListener)
> >     }
> > }
> > ```
> 
> This extension may not work as expected since type is being erased at runtime. In order to fix that class type should be passed as parameter of the function or make the function inline and use reified. I don't think inlining this function is a good idea. So I will be passing the type as a parameter similar to the `getValue(T)` in `DataSnapshot` class.
> 
> ```
> suspend fun <T : Any> Query.await(type: Class<T>): T? {
>     return suspendCancellableCoroutine { continuation ->
>         val eventListener = object : ValueEventListener {
>             override fun onDataChange(dataSnap: DataSnapshot) {
>                 continuation.resume(dataSnap.getValue(type))
>             }
> 
>             override fun onCancelled(dbError: DatabaseError) {
>                 continuation.resumeWithException(dbError.toException())
>             }
>         }
>         continuation.invokeOnCancellation {
>             removeEventListener(eventListener)
>         }
>         addListenerForSingleValueEvent(eventListener)
>     }
> }
> ```

It's an interesting suggestion @msasikanth If you could provide a PR it would be great. Please, do  keep in mind that changes in the API are subject to internal approval. Thanks!

**Comment by rosariopfernandes** [comment]
@msasikanth Please let me know when you add that extension so that I can include it on #841 :)

**Comment by msasikanth** [comment]
> @msasikanth Please let me know when you add that extension so that I can include it on #841 :)

Absolutely, once I figure how to setup the project in Android Studio properly. I am having sync and test run issues with it.

**Comment by msasikanth** [comment]
@rosariopfernandes , @rlazo I have created the PR for the extension 

https://github.com/firebase/firebase-android-sdk/pull/862

**Pull request merged**